docs/custom-filters.adoc

```diff
@@ -45,23 +151,158 @@ Instances are specific to the channel between a single client and a single broke
+> -- https://kafka.apache.org/protocol.html#protocol_compatibility
+
+You may wish to restrict your Filter to only apply to specific versions of an API. For example, "intercept all FetchRequest
+messages greater than api version 7". To do this you can override a method named `shouldHandleXyz[Request|Response]` on your filter like:
+
+[source,java]
+----
+public class FetchFilter implements FetchRequestFilter{
+
+    @Override
+    public void onFetchRequest(short apiVersion, RequestHeaderData header, FetchRequestData request, KrpcFilterContext context) {
+        context.forwardRequest(header, request);
+    }
+
+    @Override
+    public boolean shouldHandleFetchRequest(short apiVersion) {
+        return apiVersion >= 7;
+    }
+}
+----
+
 === Handling state
 
 The simplest way of managing per-client state is to use member fields.
 The proxy guarantees that all methods of a given filter instance will always be invoked on the same thread.
 Therefore there is no need to use synchronization when accessing such fields.
 
-=== An example protocol filter
+=== Filter Construction and Configuration
+:github-sample-contributor: {github}/blob/main/kroxylicious-sample/src/main/java/io/kroxylicious/sample/SampleContributor.java
 
-// TODO
+For Kroxylicious to instantiate and configure your custom filter we use Java's https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html[ServiceLoader] API.
+A Custom Filter JAR should provide a {filter-api-javadoc}/io/kroxylicious/proxy/filter/FilterContributor.html[FilterContributor] implementation that can contribute an instance of your custom Filter and
+optionally declare a configuration class that Kroxylicious will populate (using Jackson) when loading your custom Filter.
+The module must package a `META-INF/services/io.kroxylicious.proxy.filter.FilterContributor` file containing the
+classname of the contributor into the JAR file.
 
-== Packaging filters
+A convenience is provided to simplify building FilterContributors, the {api-javadoc}/io/kroxylicious/proxy/service/BaseContributor.html[BaseContributor]. See the {github-sample-contributor}[SampleContributor] for
+an example implementation (and corresponding {github}/blob/main/kroxylicious-sample/src/main/resources/META-INF/services/io.kroxylicious.proxy.filter.FilterContributor[service metadata]).
 
-Filters are packaged as standard `.jar` files.
+For simple integration with the BaseContributor a Custom Protocol Filter may have:
 
-// TODO
+1. No Constructor
+2. A no-arguments Constructor
+3. A Constructor where the only parameter is a class that extends {api-javadoc}/io/kroxylicious/proxy/config/BaseConfig.html[BaseConfig].
 
-== Sample filters
+If you use style 3, you can supply your own configuration class. This class must be deserializable using https://github.com/FasterXML/jackson[FasterXML/Jackson]
```

Comment:
We can be a little stronger here when using style three authors will generally supply their own configuration class. Otherwise why would they bother?